tests/pipeline/test_refresh_modes.py

```diff
@@ -0,0 +1,266 @@
+                assert "source_key_1" not in dlt.state()
+                assert "source_key_2" not in dlt.state()
+                assert "source_key_3" not in dlt.state()
+                resource_state("some_data_1")["resource_key_3"] = "resource_value_3"
+            yield {"id": 1, "name": "John"}
+            yield {"id": 2, "name": "Jane"}
+
+        @dlt.resource
+        def some_data_2():
+            if first_run:
+                dlt.state()["source_key_2"] = "source_value_2"
+                resource_state("some_data_2")["resource_key_3"] = "resource_value_3"
+                resource_state("some_data_2")["resource_key_4"] = "resource_value_4"
+            yield {"id": 3, "name": "Joe"}
+            yield {"id": 4, "name": "Jill"}
+
+        @dlt.resource
+        def some_data_3():
+            if first_run:
+                dlt.state()["source_key_3"] = "source_value_3"
+                resource_state("some_data_3")["resource_key_5"] = "resource_value_5"
+            yield {"id": 5, "name": "Jack"}
+            yield {"id": 6, "name": "Jill"}
+
+        @dlt.resource
+        def some_data_4():
+            yield []
+
+        return [some_data_1, some_data_2, some_data_3, some_data_4]
+
+    # First run pipeline with load to destination so tables are created
+    pipeline = dlt.pipeline(
+        "refresh_full_test",
+        destination="duckdb",
+        refresh="drop_dataset",
+        dataset_name="refresh_full_test",
+    )
+
+    info = pipeline.run(my_source())
+    assert_load_info(info)
+
+    # Second run of pipeline with only selected resources
+    first_run = False
+    info = pipeline.run(my_source().with_resources("some_data_1", "some_data_2"))
+    # pipeline.extract(my_source().with_resources("some_data_1", "some_data_2"))
+    # pipeline.normalize()
+    # pipeline.load()
+
+    assert set(t["name"] for t in pipeline.default_schema.data_tables(include_incomplete=True)) == {
+        "some_data_1",
+        "some_data_2",
+        # Table has never seen data and is not dropped
+        "some_data_4",
+    }
+
+    # Confirm resource tables not selected on second run got wiped
+    with pytest.raises(DatabaseUndefinedRelation):
+        with pipeline.sql_client() as client:
+            result = client.execute_sql("SELECT * FROM some_data_3")
+
+    with pipeline.sql_client() as client:
+        result = client.execute_sql("SELECT id FROM some_data_1 ORDER BY id")
+    assert result == [(1,), (2,)]
+
+    # Loaded state contains only keys created in second run
+    with pipeline.destination_client() as dest_client:
+        destination_state = load_pipeline_state_from_destination(
+            pipeline.pipeline_name, dest_client  # type: ignore[arg-type]
+        )
+    assert destination_state["sources"]["my_source"]["resources"] == {
+        "some_data_1": {"resource_key_3": "resource_value_3"},
+    }
+
+
+def test_refresh_drop_tables():
+    first_run = True
+
+    @dlt.source
```

Comment:
is this code by any chance fully duplicated?

Reply:
Pretty much. Cleaned this up a bit, the source is defined only once now